### ISSUE TITLE
Job Server Latency

### ISSUE THREAD

**Issue by hntd187** [spark-jobserver/spark-jobserver#403]
Hi, I have the job server running on my master, a m4.2xlarge in an EMR cluster. My jobs run, they return proper results, everything is fine with the functionality, but before when I ran jobs via spark-submit my custom jobs were taking about 20 seconds on average, now with the job server in the middle, they are taking upwards of 55-60 seconds. Perhaps i just need to give the job server some more hardware, but is there a reason for this latency addition and is it something that would scale with the job size? Right now my jobs crunch about 4-5gb of data, but the latency doubles with the job server. What information can I provide to assist in understanding this? 

I'm currently having it run with the H2 in Memory DB, but I was going to switch it over to a small RDS postgres db. Would that help at all?

**Comment by velvia** [comment]
Hmmm….  are you returning large job results as the output of the runJob method?   That is one thing that would definitely slow it down.

Maybe tell us a bit more about what your jobs do?

The other thing is to watch the configuration of your job/context, make sure they are configured the same as before with spark-submit.  Which version are you using?    It could be that the driver or executors are not getting the same amount of memory as before.

If you are using the latest job server, it forks off each context to a separate process, and there might be a bug in terms of configuring the forked process with the right amount of memory.

**Comment by hntd187** [comment]
@velvia  It's Job Server 0.6.1, and EMR 4.3, So Spark 1.6 and Hadoop 2.7

My jobs return nothing, they are a unit method, their final output is written as a parquet file to S3 that uses the EMRFS for the S3 access. 

The reason I think SJS is responsible is because the job will finish according to both the spark history server as well as the output on S3, but the job server will sit around for some amount of time before it finally flags the job as being finished. I have my jobs internally measure their run time and my jobs themselves are about 20-30 seconds longer in the internal time. My internal time and the job server times never line up, but that's to be expected, they are off normally by about 20-ish seconds, but my jobs internally are taking longer. Could perhaps the SJS need to know about native hadoop libraries or something like that? 

I've checked my configuration up and down to the point where I'm almost positive they are the same, but I might be making a mistake. The cluster itself is x4 C4.4xlarges, I've configured them with everything from having fat executors like 5cpu cores and 7G of memory to making lots of small executors like 1 cpu core and 1524m of memory. Everytime I changed the configuration, I shut down the job server, clear all it's data and then reupload an application jar, create a new context and run it. Mind you I am using an SQLContext and not a standard SparkContext, but if it'd be helpful I can show you my /usr/lib/spark/conf/spark-defaults.conf vs my emr.conf configuration in the job server.

**Comment by velvia** [comment]
Hi Stephen,

What Id like to see: 
- the config files
- the rest calls / curl commands used
- a comparison of amount of memory and cores used for executors with and without SJS

-Evan
"Never doubt that a small group of thoughtful, committed citizens can change the world" - M. Mead

**Comment by hntd187** [comment]
Here is my configuration in my emr.conf

```
spark {
  master = "yarn-client"
  jobserver {
    port = 8090
    jar-store-rootdir = /mnt/tmp/spark-jobserver/jars
    jobdao = spark.jobserver.io.JobFileDAO
    context-per-jvm = true
    filedao {
      rootdir = /mnt/tmp/spark-jobserver/filedao/data
    }
    max-jobs-per-context = 32
  }
  context-settings {
    num-cpu-cores = 4
    memory-per-node = 5g
    spark.yarn.preserve.staging.files = true
    spark.serializer = org.apache.spark.serializer.KryoSerializer
    spark.sql.parquet.output.committer.class = org.apache.spark.sql.parquet.DirectParquetOutputCommitter
    spark.sql.parquet.compression.codec = snappy
    spark.scheduler.mode = FAIR
    spark.driver.extraClassPath = "/etc/hadoop/conf:/etc/hive/conf:/usr/lib/hadoop/*:/usr/lib/hadoop-hdfs/*:/usr/lib/hadoop-yarn/*:/usr/lib/hadoop-lzo/lib/*:/usr/share/aws/aws-java-sdk/*:/usr/share/aws/emr/emrfs/conf:/usr/share/aws/emr/emrfs/lib/*:/usr/share/aws/emr/emrfs/auxlib/*"
    spark.driver.extraLibraryPath = "/usr/lib/hadoop/lib/native:/usr/lib/hadoop-lzo/lib/native"
    spark.executor.extraClassPath = "/etc/hadoop/conf:/etc/hive/conf:/usr/lib/hadoop/*:/usr/lib/hadoop-hdfs/*:/usr/lib/hadoop-yarn/*:/usr/lib/hadoop-lzo/lib/*:/usr/share/aws/aws-java-sdk/*:/usr/share/aws/emr/emrfs/conf:/usr/share/aws/emr/emrfs/lib/*:/usr/share/aws/emr/emrfs/auxlib/*"
    spark.executor.extraLibraryPath = "/usr/lib/hadoop/lib/native:/usr/lib/hadoop-lzo/lib/native"
    spark.eventLog.enabled = true
    spark.eventLog.dir = "hdfs:///var/log/spark/apps"
    spark.history.fs.logDirectory = "hdfs:///var/log/spark/apps"
    spark.yarn.historyServer.address = "ip-10-50-6-52.ec2.internal:18080"
    spark.history.ui.port = 18080
    spark.shuffle.service.enabled = true
    spark.driver.extraJavaOptions = "-Dlog4j.configuration=file:///etc/spark/conf/log4j.properties -XX:+UseConcMarkSweepGC -XX:CMSInitiatingOccupancyFraction=70 -XX:MaxHeapFreeRatio=70 -XX:+CMSClassUnloadingEnabled -XX:MaxPermSize=512M -XX:OnOutOfMemoryError='kill -9 %p'"
    spark.dynamicAllocation.enabled = true
    spark.executor.extraJavaOptions = "-verbose:gc -XX:+PrintGCDetails -XX:+PrintGCDateStamps -XX:+UseConcMarkSweepGC -XX:CMSInitiatingOccupancyFraction=70 -XX:MaxHeapFreeRatio=70 -XX:+CMSClassUnloadingEnabled -XX:OnOutOfMemoryError='kill -9 %p'"
    spark.default.parallelism = 200
    passthrough {
      spark.sql.parquet.output.committer.class = org.apache.spark.sql.parquet.DirectParquetOutputCommitter
      mapreduce.fileoutputcommitter.marksuccessfuljobs = false
    }
  }
  home = "/usr/lib/spark"
}
```
and in /usr/lib/spark/conf/spark-defaults.conf

```
spark.master                     yarn
spark.driver.extraClassPath      /etc/hadoop/conf:/etc/hive/conf:/usr/lib/hadoop/*:/usr/lib/hadoop-hdfs/*:/usr/lib/hadoop-yarn/*:/usr/lib/hadoop-lzo/lib/*:/usr/share/aws/aws-java-sdk/*:/usr/share/aws/emr/emrfs/conf:/usr/share/aws/emr/emrfs/lib/*:/usr/share/aws/emr/emrfs/auxlib/*
spark.driver.extraLibraryPath    /usr/lib/hadoop/lib/native:/usr/lib/hadoop-lzo/lib/native
spark.executor.extraClassPath    /etc/hadoop/conf:/etc/hive/conf:/usr/lib/hadoop/*:/usr/lib/hadoop-hdfs/*:/usr/lib/hadoop-yarn/*:/usr/lib/hadoop-lzo/lib/*:/usr/share/aws/aws-java-sdk/*:/usr/share/aws/emr/emrfs/conf:/usr/share/aws/emr/emrfs/lib/*:/usr/share/aws/emr/emrfs/auxlib/*
spark.executor.extraLibraryPath  /usr/lib/hadoop/lib/native:/usr/lib/hadoop-lzo/lib/native
spark.eventLog.enabled           true
spark.eventLog.dir               hdfs:///var/log/spark/apps
spark.history.fs.logDirectory    hdfs:///var/log/spark/apps
spark.yarn.historyServer.address ip-10-50-6-52.ec2.internal:18080
spark.history.ui.port            18080
spark.shuffle.service.enabled    true
spark.driver.extraJavaOptions    -Dlog4j.configuration=file:///etc/spark/conf/log4j.properties -XX:+UseConcMarkSweepGC -XX:CMSInitiatingOccupancyFraction=70 -XX:MaxHeapFreeRatio=70 -XX:+CMSClassUnloadingEnabled -XX:MaxPermSize=512M -XX:OnOutOfMemoryError='kill -9 %p'
spark.dynamicAllocation.enabled  true
spark.executor.extraJavaOptions  -verbose:gc -XX:+PrintGCDetails -XX:+PrintGCDateStamps -XX:+UseConcMarkSweepGC -XX:CMSInitiatingOccupancyFraction=70 -XX:MaxHeapFreeRatio=70 -XX:+CMSClassUnloadingEnabled -XX:OnOutOfMemoryError='kill -9 %p'
spark.executor.memory            5G
spark.executor.cores             4
spark.default.parallelism        200
```

Here are the 2 latest jobs ran through SJS
```json
/*
Job Results
curl localhost:8090/jobs | jq .[0:2]
*/
[
  {
    "duration": "79.777 secs",
    "classPath": "com.planalytics.forecasting.ForecastJob",
    "startTime": "2016-03-17T16:11:18.284Z",
    "context": "sql-context",
    "status": "FINISHED",
    "jobId": "c83a4656-7d9c-4fdc-aa49-239d9c170424"
  },
  {
    "duration": "82.373 secs",
    "classPath": "com.planalytics.forecasting.ForecastJob",
    "startTime": "2016-03-17T16:08:58.777Z",
    "context": "sql-context",
    "status": "FINISHED",
    "jobId": "e099dcf1-7c76-498e-b64c-678941ebd355"
  }
]
// Results for each call Below...
// curl localhost:8090/jobs/e099dcf1-7c76-498e-b64c-678941ebd355 | jq .
[
  {
    "duration": "79.777 secs",
    "classPath": "com.planalytics.forecasting.ForecastJob",
    "startTime": "2016-03-17T16:11:18.284Z",
    "context": "sql-context",
    "result": 65138,
    "status": "FINISHED",
    "jobId": "c83a4656-7d9c-4fdc-aa49-239d9c170424"
  },
  {
    "duration": "82.373 secs",
    "classPath": "com.planalytics.forecasting.ForecastJob",
    "startTime": "2016-03-17T16:08:58.777Z",
    "context": "sql-context",
    "result": 63704,
    "status": "FINISHED",
    "jobId": "e099dcf1-7c76-498e-b64c-678941ebd355"
  }
]
```

Collected with
```
curl localhost:8090/jobs | jq .[0:2]
curl localhost:8090/jobs/e099dcf1-7c76-498e-b64c-678941ebd355 | jq .
```

Finally my jobs ran with just spark submit

```
[root@ip-0-0-0-0~]# spark-submit --class com.planalytics.forecasting.Forecasting ingestion-appjar-2_10-0.1.5.jar
20687
[root@ip-0-0-0-0 ~]# spark-submit --class com.planalytics.forecasting.Forecasting ingestion-appjar-2_10-0.1.5.jar
23040
[root@ip-0-0-0-0~]# spark-submit --class com.planalytics.forecasting.Forecasting ingestion-appjar-2_10-0.1.5.jar
17490
```

Running jobs with any different memory or core configuration does nothing on these times. My spark submit jobs remain fast while my SJS jobs remain in the same realm of time. Hopefully this is enough information.

**Comment by velvia** [comment]
Stephen,

Many thanks.  I can’t see anything off the top of my head.   Perhaps it is the driver slowing things down.

What I would do to investigate further is to look at the Spark UI timeline and DAG views for the jobs - both the ones submitted by SJS and the regular spark-submit.   I would look to see how the timelines differ, if the DAGs are any different.

thanks,
Evan

**Comment by hntd187** [comment]
@velvia So I did a number of things.

1. I checked the DAGs, they are identical.
2. I checked the scheduling timelines they are the same.
3. I switched from EMRFS to S3A, no noticeable difference.
4. I gave the Application Master more resources, nothing.
5. I did notice that you set spark.max.cores in the config for the num-cpu-cores, which in a YARN config is not correct. This property shouldn't be set in YARN and you should set spark.executor.cores. Well that is if you want the job to request as many cores as it can get. 
6. I tripled checked the config 25 more times to make sure I wasn't crazy.
7. I profiled the Job Server during a job run, and I think I have something that seems a bit odd to me, but I am not sure if this is the cause. Across the run, at the end after the spark job has completed I notice that org.apache.hadoop.io.retry.RetryInvocationHandler shoots up in CPU. I'm not sure why. 

![screenshot](https://cloud.githubusercontent.com/assets/6778339/13968189/bc159614-f051-11e5-806e-3d1ef67dfcaf.png)

I dunno man, at this point I'm kinda out of ideas and just gonna have to deal with it or figure out another way.

**Comment by velvia** [comment]
Wait, you said the scheduling timelines are the same, however you had said that you measured the internal runtimes in your jobs and they were double in SJS.   So there must be some discrepancy between what Spark is measuring and what your job is measuring, right?

How about this …. wrap the spark jobserver job in a main, which creates its own context and then submit that in spark-submit…. what happens?

The RetryHandler seems to suggest that some I/O gets retried in the SJS case, which I don’t know why that would be.   However that would show up in the timelines…. as being different.

-Evan

**Comment by hntd187** [comment]
The actual times reported by spark are normal fast times, inline with my job running fine. The times that are reported by the SJS are wildly off in comparison.

**Comment by velvia** [comment]
Perhaps what happened is that SJS is slow in recognizing when the job is done.  However it could actually be done, in which case it is fixable on our end ….

**Comment by hntd187** [comment]
I'm speaking purely from empirically watching it, but it seems like one of two things is occurring. The job finishes and SJS is slow picking up on the finished state or the Spark cluster itself is slow to notify the SJS that the job has been completed and write back to the driver the results.

**Comment by velvia** [comment]
Well, what I can do (or you can do) is to instrument this part of the SJS code:

      try {
        statusActor ! JobStatusActor.JobInit(jobInfo)

        val jobC = jobContext.asInstanceOf[job.C]
        job.validate(jobC, jobConfig) match {
          case SparkJobInvalid(reason) => {
            val err = new Throwable(reason)
            statusActor ! JobValidationFailed(jobId, DateTime.now(), err)
            throw err
          }
          case SparkJobValid => {
            statusActor ! JobStarted(jobId: String, contextName, jobInfo.startTime)
            val sc = jobContext.sparkContext
            sc.setJobGroup(jobId, s"Job group for $jobId and spark context ${sc.applicationId}", true)
            job.runJob(jobC, jobConfig)
          }
        }
      } finally {
        org.slf4j.MDC.remove("jobId")
      }
    }(executionContext).andThen {
      case Success(result: Any) =>
        statusActor ! JobFinished(jobId, DateTime.now())
        resultActor ! JobResult(jobId, result)
      case Failure(error: Throwable) =>
        // If and only if job validation fails, JobErroredOut message is dropped silently in JobStatusActor.
        statusActor ! JobErroredOut(jobId, DateTime.now(), error)
        logger.warn("Exception from job " + jobId + ": ", error)


As you can see, once the runJob finishes, then the future callback is invoked to send the JobFinished message.

Thus a couple things could be happening:
1) Your job spends a lot of time in the validate() method.  The runJob itself isn’t any longer.   From your empirical observation this might not be true, but you can validate that by calling the async API, which calls validate.  Also your job would be in the started state.
2) For some reason, the callback that sends JobFinished is not scheduled immediately after the job finishes.   

2) could happen if you are running simultaneous jobs or somehow the thread pool is too small.

**Comment by velvia** [comment]
@hntd187 hey Steve, is this still happening?  What do you think of the thoughts above?

**Comment by hntd187** [comment]
I did what you had suggested and I didn't find anything that jumped out to me off it. I figured it had to be with the writing of the parquet files back to S3, which is what eventually lead to my hadoop configuration pull request, but I haven't gotten a significant time to test it on large jobs, I'll probably be doing that sometime later this week or early next week. If you want to close this we can be done with it. If the parquetdirectoutputcommitter isn't the problem then I'm just gonna accept it.
